pkg/scheduler/controller/shoot/reconciler.go

```diff
@@ -26,46 +26,45 @@ import (
 	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
 	cidrvalidation "github.com/gardener/gardener/pkg/utils/validation/cidr"
 
-	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/validation/field"
 	"k8s.io/client-go/tools/record"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
 // MsgUnschedulable is the Message for the Event on a Shoot that the Scheduler creates in case it cannot schedule the Shoot to any Seed
 const MsgUnschedulable = "Failed to schedule shoot"
 
 type reconciler struct {
-	logger       logr.Logger
 	config       *config.ShootSchedulerConfiguration
 	gardenClient client.Client
 	recorder     record.EventRecorder
 }
 
 func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
```

Comment:
How are we sure the log retrieved from the context is not nil, that is, where is the matching `IntoContext`? I only see `IntoContext` being used in `CreateWorker`, but it seems the scheduler is not using that function? I assume we are relying on some `controller-runtime` magic that always injects a logger in worker contexts?

Reply:
The counterpart is in the controller-runtime controller implementation: https://github.com/gardener/gardener/blob/8abb827821db42806a828c66b93995e1a54c693c/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go#L112-L113

The scheduler already uses this code and the other components will use this as well once they migrated to the manager. `CreateWorker` now does the same to make it easy to switch later on without thinking about logging.

BTW, `logf.FromContext` will never return `nil`. If the context is `nil` or doesn't carry a logger, the func will simply return the global `logf.Log` (which is also never `nil`, but might be a `NullLogger` if not initialized (e.g. in tests)).